Server concepts in learn (#903)
modelcontextprotocol/modelcontextprotocol: docs/docs/learn/server-concepts.mdx

| **Tools**      | Lets AI take actions for you (with permission) | Model-controlled       | Search flights, send messages, create calendar events        |

Let's explore each of these using our vacation planning example to see how they work together.

## Resources: Giving AI Access to Information

Resources are how MCP lets AI assistants understand context without you having to copy and paste everything manually. They're like giving the AI "read access" to relevant information.

**In our vacation example**, resources might include:

- Your calendar (to check when you're free)
- Weather forecasts for destinations
- Travel blogs and guides you've saved
- Your previous trip itineraries
- Restaurant recommendations from friends

### How Resources Work

Instead of manually telling the AI "I'm free from June 15-22, and last year I went to Barcelona, and my friend recommended this great restaurant in Paris...", resources let the AI access this information directly (with your permission, of course).

Resources can be:

- **Static**: Specific files or data (like "Barcelona Trip Notes.pdf")

> - **Dynamic**: Patterns that fetch information on demand (like "weather://forecast/{city}/{date}" for any city's weather)

Author: updated

**Protocol operations:**

| Method       | Purpose                  | Returns                                |
| ------------ | ------------------------ | -------------------------------------- |
| `tools/list` | Discover available tools | Array of tool definitions with schemas |
| `tools/call` | Execute a specific tool  | Tool execution result                  |

**Example tool definition:**

```typescript
{
  name: "searchFlights",
  description: "Search for available flights",
  inputSchema: {
    type: "object",
    properties: {
      origin: { type: "string", description: "Departure city" },
      destination: { type: "string", description: "Arrival city" },
      date: { type: "string", format: "date", description: "Travel date" }
    },
    required: ["origin", "destination", "date"]
  }
}
```

#### Example: Taking Action

Tools transform AI from an advisor into an assistant that can help execute tasks. In a travel planning scenario, the AI application might orchestrate several actions to book a vacation.

> First, it searches for flights using `searchFlights(origin: "NYC", destination: "Barcelona", date: "2024-06-15")`, which queries multiple airlines and returns structured flight options. Once flights are selected, it creates a calendar event with `createCalendarEvent(title: "Barcelona Trip", startDate: "2024-06-15", endDate: "2024-06-22")` to block the travel dates. Finally, it sends an out-of-office notification using `sendEmail(to: "team@work.com", subject: "Out of Office", body: "...")` to inform colleagues about the absence.

Author: yeah, but let's keep it one style - updating, thanks

**Protocol operations:**

| Method                     | Purpose                         | Returns                                |
| -------------------------- | ------------------------------- | -------------------------------------- |
| `resources/list`           | List available direct resources | Array of resource descriptors          |
| `resources/templates/list` | Discover resource templates     | Array of resource template definitions |
| `resources/read`           | Retrieve resource contents      | Resource data with metadata            |
| `resources/subscribe`      | Monitor resource changes        | Subscription confirmation              |

#### Example: Accessing Context Data

Continuing with the travel planning example, resources provide the AI application with access to relevant information:

- **Calendar data** (`calendar://events/2024`) - To check availability
- **Weather forecasts** (`weather://forecast/{city}/{date}`) - For destination climate
- **Travel documents** (`file:///Documents/Travel/passport.pdf`) - For important information
- **Previous itineraries** (`trips://history/barcelona-2023`) - For reference

> Instead of manually copying this information, resources let the AI access it directly with user permission. During planning, the AI can check calendar availability, look up weather patterns, and reference travel preferences—all through structured resource access.

Author: "Tools vs Resources: how to choose" I'll do it as a separate PR

**Resource Template Examples:**

```json
{
  "uriTemplate": "weather://forecast/{city}/{date}",
  "name": "weather-forecast",
  "title": "Weather Forecast",
  "description": "Get weather forecast for any city and date",
  "mimeType": "application/json"
}

{
  "uriTemplate": "travel://flights/{origin}/{destination}",
  "name": "flight-search",
  "title": "Flight Search",
  "description": "Search available flights between cities",
  "mimeType": "application/json"
}
```

These templates enable flexible queries. For weather data, users can access forecasts for any city/date combination. For flights, they can search routes between any two airports. When users type "travel://flights/NYC/Bar", the system can suggest "Barcelona (BCN)" or "Barbados (BGI)" for the destination parameter.

> #### Parameter Discovery

Reviewer: "Discovery" or "Completion"?

```suggestion
#### Parameter Completion
```

Dynamic resources support parameter completion. For example:

> - Typing "Par..." in `weather://forecast/{city}` might suggest "Paris" or "Park City"

Reviewer: ```suggestion
- Typing "Par" as input for `weather://forecast/{city}` might suggest "Paris" or "Park City"
```

### Prompts - interaction templates

Prompts provide reusable templates that guide interactions. They combine instructions with parameter definitions to standardize common workflows.

#### Overview

Prompts are structured templates that define expected inputs and interaction patterns. They are user-controlled, requiring explicit invocation rather than automatic triggering. Prompts can be context-aware, referencing available resources and tools to create comprehensive workflows.

**Protocol operations:**

| Method         | Purpose                    | Returns                               |
| -------------- | -------------------------- | ------------------------------------- |
| `prompts/list` | Discover available prompts | Array of prompt descriptors           |
| `prompts/get`  | Retrieve prompt details    | Full prompt definition with arguments |

**Prompt components:**

- **Name**: Unique identifier for the prompt
- **Description**: User-facing explanation of purpose
- **Arguments**: Typed parameters with optional defaults

> - **Instructions**: Template text with parameter placeholders

Reviewer: Regarding "Instructions", the server doesn't expose the template text to the client, and the actual prompt content that the server returns is termed "messages" (e.g. `messages[i].content`).  What do we want to highlight here?

Also, similar to previous comments, I think `title` is currently preferred over `name`, and these fields may be optional.

Author: name is unique identifier, where title is preferred for a display. I think I will just remove this section, it's not super useful here

#### Example: Streamlined Workflows

Prompts provide structured templates for common tasks. In the travel planning context:

**"Plan a vacation" prompt:**

```yaml
name: plan-vacation
description: Guide through vacation planning process
arguments:
  - destination: string (required)
  - duration: number (days)
  - budget: number (optional)
  - interests: array of strings

> ```

Reviewer: Why YAML here?

Author: just for the formatting

The protocol gives implementers freedom to design interfaces that feel natural within their application. Key principles include easy discovery of available prompts, clear descriptions of what each prompt does, natural argument input with validation, and transparent display of the prompt's underlying template.

## How It All Works Together

The real power of MCP emerges when multiple servers work together, combining their specialized capabilities through a unified interface.

### Example: Multi-Server Travel Planning

Consider an AI application with three connected servers:

1. **Travel Server** - Handles flights, hotels, and itineraries
2. **Weather Server** - Provides climate data and forecasts
3. **Calendar/Email Server** - Manages schedules and communications

#### The Complete Flow

**1. User invokes a prompt with parameters:**

```yaml
prompt: plan-vacation
arguments:
  destination: "Barcelona"
  departure_date: "2024-06-15"
  return_date: "2024-06-22"
  budget: 3000
  travelers: 2

> ```

Reviewer: Why YAML here as well?

**2. User selects resources to include:**

- `calendar://my-calendar/June-2024` (from Calendar Server)
- `travel://preferences/europe` (from Travel Server)
- `travel://past-trips/Spain-2023` (from Travel Server)

> - `weather://forecast/Barcelona/2024-06` (from Weather Server)

Reviewer 2: Apropos of the resources vs tools comment above (and the bigger question of how resources are really used) - is this realistic? It seems like "I'm thinking of planning a vacation, idk Barcelona? What's the weather like there rn" in a chatbot (+ model-driven tool calls to e.g. `get_weather`) is the way it would actually be done, rather than picking city out of an app modal and typing in a date.

TBH I can't think of a good use "typical" case for parameterized resources, they all turn into tools in my mind - it's because for examples like weather and calendar, "app-controlled" implies it's the *user* doing the parameterization via app UI, which seems cumbersome compared to the model translating natural language into parameters. (For direct resources it's easier, e.g. here's my passport.)

> **1. User invokes a prompt with parameters:**
> 
> ```yaml
> prompt: plan-vacation
> arguments:
>   destination: "Barcelona"
>   departure_date: "2024-06-15"
>   return_date: "2024-06-22"
>   budget: 3000
>   travelers: 2
> ```
> 
> **2. User selects resources to include:**
> 
> - `calendar://my-calendar/June-2024` (from Calendar Server)
> - `travel://preferences/europe` (from Travel Server)
> - `travel://past-trips/Spain-2023` (from Travel Server)
> - `weather://forecast/Barcelona/2024-06` (from Weather Server)
> 
> **3. AI processes the request:**
> 
> The AI first reads all selected resources to gather context. From the calendar, it identifies available dates. From travel preferences, it learns preferred airlines and hotel types. From past trips, it discovers previously enjoyed locations. From weather data, it checks climate conditions for the travel period.
> 
> Using this context, the AI then requests user approval to execute a series of coordinated actions: searching for flights from NYC to Barcelona, finding hotels within the specified budget, creating a calendar event for the trip duration, and sending confirmation emails with the trip details.

Reviewer: I'm not sure if it was a stylistic choice, but if we want this rendered as a list with proper indentation, then we need:

```suggestion
1. **User invokes a prompt with parameters:**

    ```yaml
    prompt: plan-vacation
    arguments:
      destination: "Barcelona"
      departure_date: "2024-06-15"
      return_date: "2024-06-22"
      budget: 3000
      travelers: 2
    ```

2. **User selects resources to include:**

    - `calendar://my-calendar/June-2024` (from Calendar Server)
    - `travel://preferences/europe` (from Travel Server)
    - `travel://past-trips/Spain-2023` (from Travel Server)
    - `weather://forecast/Barcelona/2024-06` (from Weather Server)

3. **AI processes the request:**

    The AI first reads all selected resources to gather context. From the calendar, it identifies available dates. From travel preferences, it learns preferred airlines and hotel types. From past trips, it discovers previously enjoyed locations. From weather data, it checks climate conditions for the travel period.

    Using this context, the AI then requests user approval to execute a series of coordinated actions: searching for flights from NYC to Barcelona, finding hotels within the specified budget, creating a calendar event for the trip duration, and sending confirmation emails with the trip details.
```
